File: port/RangeController.py
import Action

class RangeController:
    # Constants representing up, down, left, right
    DX = [+1, 0, -1, 0]
    DY = [0, -1, 0, +1]

    @staticmethod
    def get_reachable_cells_matrix(op_unit, map):
        """Function that returns a bool matrix of the movable range of opUnit.
        Enemy units cannot be passed through, but allied units can (but cannot be the final destination)."""
        unit_spec = op_unit.get_spec()  # Spec of the unit to be moved
        unit_step = unit_spec.get_unit_step()  # Number of steps
        unit_color = op_unit.get_team_color()  # Team of the unit to be moved

        reachable = [[False] * map.get_y_size() for _ in range(map.get_x_size())]  # Reachable or not. This is returned.

        pos_counts = [0] * (unit_step + 1)  # Number of movable positions with remaining movement power as index.
        pos_x = [[0] * 100 for _ in range(unit_step + 1)]  # Movable positions with remaining movement power as index.
        pos_y = [[0] * 100 for _ in range(unit_step + 1)]  # 100 is an arbitrary number, should be made a constant.

        # The current position is a movable position with remaining movement power unitStep and is also ultimately movable.
        pos_counts[unit_step] = 1
        pos_x[unit_step][0] = op_unit.get_x_pos()
        pos_y[unit_step][0] = op_unit.get_y_pos()
        reachable[op_unit.get_x_pos()][op_unit.get_y_pos()] = True

        # From movable positions with remaining movement power restStep, look at their up, down, left, right,
        # and add them to the list if they can be moved to.
        for rest_step in range(unit_step, 0, -1):  # Current remaining movement cost
            for i in range(pos_counts[rest_step]):
                x = pos_x[rest_step][i]  # Position to focus on
                y = pos_y[rest_step][i]

                # Look at the up, down, left, right of this (x, y)
                for r in range(4):
                    new_x = x + RangeController.DX[r]  # Up, down, left, right
                    new_y = y + RangeController.DY[r]

                    new_rest = rest_step - unit_spec.get_move_cost(map.get_field_type(new_x, new_y))  # Remaining movement cost after moving
                    if new_rest < 0:
                        continue  # Hit the surroundings or not enough movement power → Cannot enter

                    u = map.get_unit_at(new_x, new_y)
                    if u is not None:
                        if u.get_team_color() != unit_color:
                            continue  # Hit an enemy unit → Cannot enter

                    # If it's not a position that has already been marked as movable, add it to the movable positions.
                    if not reachable[new_x][new_y]:
                        pos_x[new_rest][pos_counts[new_rest]] = new_x
                        pos_y[new_rest][pos_counts[new_rest]] = new_y
                        pos_counts[new_rest] += 1
                        reachable[new_x][new_y] = True

        # Positions occupied by allied units are ultimately not movable positions.
        for u in map.get_units_list(unit_color, True, True, False):
            reachable[u.get_x_pos()][u.get_y_pos()] = False

        # It's up to you how to consider your own position, but if you want to make it movable:
        reachable[op_unit.get_x_pos()][op_unit.get_y_pos()] = True

        return reachable

    @staticmethod
    def get_attack_action_list(op_unit, map):
        """Returns a list of opUnit's attack actions."""
        actions = []

        if op_unit.get_spec().is_direct_attack_type():  # If opUnit is a melee attack type
            op_units_movable_range = RangeController.get_reachable_cells_matrix(op_unit, map)  # opUnit's movable range array

            # Check if enemy units are adjacent to the movable range
            for en_unit in map.get_units_list(op_unit.get_team_color(), False, False, True):
                pos_x = en_unit.get_x_pos()  # Position of the enemy unit
                pos_y = en_unit.get_y_pos()

                for i in range(4):  # Look at the up, down, left, right of the enemy unit
                    check_x = pos_x + RangeController.DX[i]
                    check_y = pos_y + RangeController.DY[i]

                    if op_units_movable_range[check_x][check_y]:  # It means you can come to the position checkX, checkY and attack
                        actions.append(Action.create_attack_action(op_unit, check_x, check_y, en_unit))
        else:  # If it's an indirect attack type
            u_spec = op_unit.get_spec()
            # Check if enemy units are within the attackable range
            for en_unit in map.get_units_list(op_unit.get_team_color(), False, False, True):
                pos_x = en_unit.get_x_pos()  # Position of the enemy unit
                pos_y = en_unit.get_y_pos()
                dist = abs(pos_x - op_unit.get_x_pos()) + abs(pos_y - op_unit.get_y_pos())  # Distance between the enemy unit and yourself
                if u_spec.get_unit_min_attack_range() <= dist <= u_spec.get_unit_max_attack_range():
                    actions.append(Action.create_attack_action(op_unit, op_unit.get_x_pos(), op_unit.get_y_pos(), en_unit))  # Add to the list if it's within the attackable range

        return actions

    @staticmethod
    def get_attackable_cells_matrix(op_unit, map):
        """Returns the attackable range of opUnit as a bool[,] matrix.
        True is assigned to the attackable positions."""
        attackable = [[False] * map.get_y_size() for _ in range(map.get_x_size())]  # Attackable range matrix

        if op_unit.get_spec().is_direct_attack_type():  # If opUnit is a melee attack type
            movable_cells_matrix = RangeController.get_reachable_cells_matrix(op_unit, map)  # opUnit's movable range
            cells = [[False] * map.get_y_size() for _ in range(map.get_x_size())]

            for x in range(1, map.get_x_size() - 1):
                for y in range(1, map.get_y_size() - 1):
                    if map.get_unit_at(x, y) is not None and \
                            map.get_unit_at(x, y).get_team_color() == op_unit.get_team_color():
                        continue

                    # Look at the up, down, left, right positions of x, y, and positions where true is assigned are attackable range
                    for i in range(4):  # Look at the up, down, left, right of the enemy unit
                        check_x = x + RangeController.DX[i]
                        check_y = y + RangeController.DY[i]

                        if movable_cells_matrix[check_x][check_y]:
                            cells[x][y] = True  # Change to true as attackable range

            attackable = cells
        else:  # If it's an indirect attack type
            u_spec = op_unit.get_spec()

            for x in range(1, map.get_x_size() - 1):
                for y in range(1, map.get_y_size() - 1):
                    dist = abs(x - op_unit.get_x_pos()) + abs(y - op_unit.get_y_pos())  # Distance between each (x, y) coordinate and yourself
                    if u_spec.get_unit_min_attack_range() <= dist <= u_spec.get_unit_max_attack_range():  # If it's within the attackable range
                        attackable[x][y] = True  # Positions within the attackable range are set to true

        return attackable

    @staticmethod
    def is_effective(operation_unit, target_unit):
        """Check if operationUnit can attack targetUnit. Returns false if the damage effect is 0."""
        if operation_unit.get_spec().get_unit_atk_power(target_unit.get_type_of_unit()) != 0:
            return True
        return False

File: RangeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleWars 
{
    /// <summary>
    /// 移動可能範囲，攻撃可能ユニットなどを取得できるクラス．
  /// システムも使うが，思考ルーチンでの利用も可能．
  /// AI_template や AI_Sample_MaxActEval もご覧ください
    /// </summary>
    class RangeController {
        // 上下左右を表す定数
        private static int[] DX = new int[4] { +1, 0, -1, 0 };
        private static int[] DY = new int[4] { 0, -1, 0, +1 };

        // opUnitが移動できる範囲を bool の matrix で返す関数
        // 敵ユニットはすり抜けられないが，味方ユニットはすり抜けられる（最終移動先にはできない）
        public static bool[,] getReachableCellsMatrix(Unit opUnit, Map map) {
            Spec unitSpec = opUnit.getSpec();  // 動かすユニットのSpec
            int unitStep = unitSpec.getUnitStep(); // ステップ数
            int unitColor = opUnit.getTeamColor(); // 動かすユニットのチーム

            bool[,] reachable = new bool[map.getXsize(), map.getYsize()];  // 到達可能か．これを返す．boolで初期化される

            int[] posCounts = new int[unitStep + 1];  // 残り移動力 index になる移動可能箇所数．0で初期化される
            int[,] posX = new int[unitStep + 1, 100]; // 残り移動力 index になる移動可能箇所． 100は適当な数，本来 Const化すべき
            int[,] posY = new int[unitStep + 1, 100];

            // 今いる場所は残り移動力 unitStep の移動可能箇所であり，最終的にも移動可能
            posCounts[unitStep] = 1;
            posX[unitStep, 0] = opUnit.getXpos();
            posY[unitStep, 0] = opUnit.getYpos();
            reachable[opUnit.getXpos(), opUnit.getYpos()] = true;

            // 残り移動力 restStep になる移動可能箇所から，その上下左右を見て，移動できるならリストに追加していく
            for (int restStep = unitStep; restStep > 0; restStep--) { // 現在の残り移動コスト
                for (int i = 0; i < posCounts[restStep]; i++) {
                    int x = posX[restStep, i];  // 注目する場所
                    int y = posY[restStep, i];

                    // この (x,y) の上下左右を見る
                    for (int r = 0; r < 4; r++) {
                        int newx = x + DX[r]; // 上下左右
                        int newy = y + DY[r];

                        int newrest = restStep - unitSpec.getMoveCost(map.getFieldType(newx, newy)); // 移動後の残り移動コスト
                        if (newrest < 0) continue;  // 周囲にあたったか，移動力が足りない →進入不可

                        Unit u = map.getUnit(newx, newy);
                        if (u != null) {
                            if (u.getTeamColor() != unitColor) continue;  // 敵ユニットにあたった →進入不可
                        }

                        // すでに移動可能マークがついた場所じゃなければ，移動可能箇所に追加する
                        if (reachable[newx, newy] == false) {
                            posX[newrest, posCounts[newrest]] = newx;
                            posY[newrest, posCounts[newrest]] = newy;
                            posCounts[newrest]++;
                            reachable[newx, newy] = true;
                        }
                    }
                }
            }

            // 味方ユニットがいる場所は，最終的には移動可能な箇所ではない
            foreach (Unit u in map.getUnitsList(unitColor, true, true, false)) {
                reachable[u.getXpos(), u.getYpos()] = false;
            }

            // 自分の場所をどう考えるかは自由だが，移動可能とするなら
            reachable[opUnit.getXpos(), opUnit.getYpos()] = true;

            return reachable;
        }

        // opUnitの攻撃行動のリストを返す
        public static List<Action> getAttackActionList(Unit opUnit, Map map) {
            List<Action> actions = new List<Action>();

            if (opUnit.getSpec().isDirectAttackType() == true) {  // opUnitが近接攻撃タイプだった場合
                bool[,] opUnitsMovableRange = getReachableCellsMatrix(opUnit, map);// opUnitの移動可能範囲配列

                // 敵ユニットが移動可能範囲に隣接しているかどうか見る
                foreach (Unit enUnit in map.getUnitsList(opUnit.getTeamColor(), false, false, true)) {
                    int posX = enUnit.getXpos();// 敵ユニットの位置
                    int posY = enUnit.getYpos();

                    for (int i = 0; i < 4; i++) {// ここで敵ユニットの上下左右をみる
                        int checkX = posX + DX[i];
                        int checkY = posY + DY[i];

                        if (opUnitsMovableRange[checkX, checkY]) {  //checkX, checkY の位置に来て攻撃ができるということ
                            actions.Add(Action.createAttackAction(opUnit, checkX, checkY, enUnit));
                        }
                    }
                }
            } else { // 間接攻撃タイプだった場合
                Spec uSpec = opUnit.getSpec();
                // 敵ユニットが攻撃可能範囲に入っているか否か見る
                foreach (Unit enUnit in map.getUnitsList(opUnit.getTeamColor(), false, false, true)) {
                    int posX = enUnit.getXpos();// 敵ユニットの位置
                    int posY = enUnit.getYpos();
                    int dist = Math.Abs(posX - opUnit.getXpos()) + Math.Abs(posY - opUnit.getYpos()); //敵ユニットと自分との距離
                    if (uSpec.getUnitMinAttackRange() <= dist && dist <= uSpec.getUnitMaxAttackRange()) {
                        actions.Add(Action.createAttackAction(opUnit, opUnit.getXpos(), opUnit.getYpos(), enUnit));// 攻撃可能範囲に含まれるならリストに追加する
                        //actions.Add(new Action(opUnit, enUnit.ge));
                    }
                }
            }
            return actions;
        }

        // opUnitの攻撃可能範囲をbool[,] matrixで返す
        // 攻撃可能な場所にはtrueが代入されている
        public static bool[,] getAttackableCellsMatrix(Unit opUnit, Map map) {
            bool[,] attackAble = new bool[map.getXsize(), map.getYsize()];  // 攻撃可能範囲をmatrix

            if (opUnit.getSpec().isDirectAttackType()) {// opUnitが近接攻撃タイプだった場合
                bool[,] movableCellsMatrix = getReachableCellsMatrix(opUnit, map);//opUnitの移動可能範囲
                bool[,] cells = new bool[map.getXsize(), map.getYsize()];

                for (int x = 1; x < map.getXsize() - 1; x++) {
                    for (int y = 1; y < map.getYsize() - 1; y++) {
                        if (map.getUnit(x, y) != null &&
                            map.getUnit(x, y).getTeamColor() == opUnit.getTeamColor()) continue;

                        // x,yの上下左右の位置を見て，そこにtrueが代入された場所が攻撃可能範囲
                        for (int i = 0; i < 4; i++) {// ここで敵ユニットの上下左右をみる
                            int checkX = x + DX[i];
                            int checkY = y + DY[i];

                            if (movableCellsMatrix[checkX, checkY]) {
                                cells[x, y] = true;// 攻撃可能範囲としてtrueに書き換え
                            }
                        }
                    }
                }

                attackAble = cells;
            } else {// 間接攻撃タイプだった場合

                Spec uSpec = opUnit.getSpec();

                for (int x = 1; x < map.getXsize() - 1; x++) {
                    for (int y = 1; y < map.getYsize() - 1; y++) {
                        int dist = Math.Abs(x - opUnit.getXpos()) + Math.Abs(y - opUnit.getYpos()); // 各(x,y)座標と自分との距離
                        if (uSpec.getUnitMinAttackRange() <= dist && uSpec.getUnitMaxAttackRange() >= dist) {//攻撃可能範囲であれば
                            attackAble[x, y] = true;// 攻撃可能範囲に入ってる位置はtrueにする
                        }
                    }
                }
            }

            return attackAble;
        }

        // operationUnitがtargetUnitに攻撃可能かどうか．ダメージ効果0の場合にはfalseが返る
        private static bool isEffective(Unit operationUnit, Unit targetUnit) {
            if (operationUnit.getSpec().getUnitAtkPower(targetUnit.getTypeOfUnit()) != 0) { return true; }
            return false;
        }
    }
}
